### ISSUE TITLE
Too much pool management logic

### ISSUE THREAD

**Issue by NikolayPianikov** [chechulnikov/eucalypt#9]
Если использовать изоляцию на уровне контейнеров для запуска каждого скрипта, то имеет смысл добавить парамет "--rm" чтобы гаратировать утилизацию контейнера после завершения сеанса, тогда "pool" и все вокруг как "ExecutorsPoolGarbageCollector" будут не нужны. Я бы просто добавил декоратор над Executor, который внутри себя ограничивал "пропускную способность" используя "semaphore"

**Comment by chechulnikov** [comment]
Изначально исходил из посыла, что docker run – это может быть дорого и надо держать часть контейнеров запущенными.

Согласен, что упростить можно. Но пул решает ещё две задачи:
- поддержка некоторого количества постоянно запущенных и быстро доступных к запуску скрипта контейнеров
- автоматическое расширение при оказании давления на систему
- автоматические уменьшение размера, если количество постоянно запущенных контейнеров слишком велико

> Я бы просто добавил декоратор над Executor, который внутри себя ограничивал использование ресурсов через "semaphore"

Боюсь, в предлагаемом варианте контейнеры придётся постоянно стартовать. Сделать так, значит полагаться всё время на "холодный" старт

**Comment by NikolayPianikov** [comment]
Старт контейнера довольно быстрый если образ уже скачан и это не первый старт

**Comment by chechulnikov** [comment]
Я специально оставил такую возможность, потому что я не знаю, какие контейнеры у меня будут в будущем – вдруг там старт какого-нибудь тяжелого демона при старте (ENTRYPOINT/CMD контейнера) должен быть и его нельзя сделать при билде, а только при docker run?

Сейчас мы ничего не знаем о потенциально добавляющихся новых executor-ах.

**Comment by NikolayPianikov** [comment]
> Старт может быть и не быстрым.

я согласен, но сейчас вы по сути реализовали не pool а некий механизм который предварительно "разогревает" создает 3 контейнера заранее, при этом эти контейнеры переиспользовать фактически нельзя, а это основная задача pool

**Comment by chechulnikov** [comment]
Пул executor-ов, а не контейнеров. Переиспользуются executor-ы, а не контейнеры. То что executor представляет из себя контейнер – деталь реализации и проблемы самого executor-а. Вы можете переиспользовать вместо контейнеров пользователей с ограниченными правами, как вы предлагали в соседнем issue переписав реализацию `Executor`.

Возможно, я где-то писал слова "пул контейнеров", так то чтобы идея была понятнее. Прошу прощения, если запутал.


> не pool а некий механизм 

> переиспользовать фактически нельзя, а это основная задача pool

Давайте не будем называть это пулом, а назовём просто менеджером (что не оч понятно) или как-то иначе.
